### DISCUSSION TITLE
#1463 request stream的一个小问题

### DISCUSSION THREAD

**Discussion by yxhtx** [cloudwu/skynet#1611] [General]
我的代码：
```
skynet.fork(function()
	while true do
		local _stream<close> = stream
		local succ, _result = pcall(read_watch_stream, self, _stream)
		if not succ then
			skynet.logerror("read_watch_stream", _result)
			break
		end
		skynet.fork(watchcb, self, _result)
	end
end)
```
依赖to-be-closed特性自动调用stream.close时，最后会wakeup dead coroutine：
```
-- SKYNET_SOCKET_TYPE_CLOSE = 3
socket_message[3] = function(id)
	local s = socket_pool[id]
	if s == nil then
		return
	end
	s.connected = false
	wakeup(s)
end
```
请问是否应该判断dead croutine，还是说从使用上避免？

**Comment by cloudwu** [comment]
我没看明白你到底用 to be closed 做什么，期望在什么时候调用它。

**Reply by yxhtx** [reply]
我使用httpc.request_stream返回的stream对象来循环读取数据直至not succ，这时fork的函数将会结束，期望__close能被自动调用。
不过其实我也能手动调用一下stream:close()也代替这种操作。

**Reply by cloudwu** [reply]
那么这个问题到底是 __close 引起的还是别的原因引起的呢？你的代码没看出 to be closed 的必要性。

**Reply by cloudwu** [reply]
这里为什么会 wakeup dead coroutine ? 为什么 not succ 时，对应的 socket 对象中的 s.co 不为空？

**Reply by yxhtx** [reply]
是 __close 引起的
我是基于以下两点想这样用的：
1.网络出错succ返回false时，stream自动close；
2.调用skynet.abort时，回收虚拟机时，这些stream也能自动close。

**Reply by yxhtx** [reply]
貌似是因为socket.close(id)的suspend(s)里给s.co赋值了

**Reply by cloudwu** [reply]
我明白了。这个问题是目前 lua 不支持 yield from __close 引起的。 http://lua-users.org/lists/lua-l/2020-10/msg00009.html

常规 socket.close() 会 yield coroutine，等待底层的 SOCKET CLOSE 消息，然后才结束当前 coroutine 。而 __close 跳过了这个环节。如果下一个版本 lua 支持了 yield from __close 就自然没有问题；目前我还需要想个方法绕过这个问题。

**Reply by cloudwu** [reply]
Lua 5.4.4 已经支持了 yield form __close, 你是不是没更新仓库呢？

**Reply by yxhtx** [reply]
好的，我的lua版本还是5.4.2的，那我更新一下。
